File: src/Dailies/September/15-MaxNumberOfWords/MaxNumberOfWords.cs
namespace LeetCode.Dailies.MaxNumberOfWords;

/// <summary>
/// There is a malfunctioning keyboard where some letter keys do not work. All other keys on the keyboard work properly.
/// Given a string text of words separated by a single space(no leading or trailing spaces) and a string brokenLetters of 
/// all distinct letter keys that are broken, return the number of words in text you can fully type using this keyboard.
/// 
/// Constraints
/// 1 <= text.length <= 10^4
/// 0 <= brokenLetters.length <= 26
/// text consists of words separated by a single space without any leading or trailing spaces.
/// Each word only consists of lowercase English letters.
/// brokenLetters consists of distinct lowercase English letters.
/// 
/// https://leetcode.com/problems/maximum-number-of-words-you-can-type
/// </summary>
public class Solution
{
    public int CanBeTypedWords(string text, string brokenLetters)
    {
        HashSet<char> brokens = [.. brokenLetters];
        int words = 0;
        bool skip = false;
        for (int i = 0; i < text.Length; i++)
        {
            char letter = text[i];
            if (brokens.Contains(letter))
            {
                skip = true;
                continue;
            }
            if (text[i] == ' ' || i == text.Length - 1)
            {
                words = skip ? words : words + 1;
                skip = false;
            }
        }

        return words;
    }
}

File: src/GoAlgos/Dailies/September/15-max_number_of_words/max_number_of_words.go
package godailies

func canBeTypedWords(text string, brokenLetters string) int {
	var seen [26]bool
	for _, r := range brokenLetters {
		if 'a' <= r && r <= 'z' {
			seen[r-'a'] = true
		}
	}

	skip := false
	words := 0
	for i, c := range text {
		if c != ' ' && seen[c-'a'] {
			skip = true
			continue
		}
		if c == ' ' || i == len(text)-1 {
			if !skip {
				words++
			}
			skip = false
		}
	}

	return words
}
